book/forms.md

```diff
@@ -261,119 +261,104 @@ slow in your browser, so typing into input forms is actually going to
     browser implements that.
 
-``` {.python}
-def submit_form(self, elt):
-    while elt and elt.tag != "form":
+``` {.python indent=12}
+elif elt.tag == "button":
+    while elt:
+        if elt.tag == "form" and "action" in elt.attributes:
+            return self.submit_form(elt)
         elt = elt.parent
-    if not elt: return
 ```
 
-Fourth, we need to find all of the input elements inside this form:
+The `submit_form` method is then in charge of finding all of the input
+elements, encoding them in the right way, and sending the `POST`
+request. First, we look through all the descendents of the `form` to
+find `input` elements:
 
 ``` {.python}
-def find_inputs(elt, out):
-    if not isinstance(elt, ElementNode): return
-    if elt.tag == "input" and "name" in elt.attributes:
-        out.append(elt)
-    for child in elt.children:
-        find_inputs(child, out)
-    return out
+class Tab:
+    def submit_form(self, elt):
+        inputs = [node for node in tree_to_list(elt)
+                  if isinstance(node, Element)
+                  and node.tag == "input"
+                  and "name" in node.attributes]
 ```
 
-Fifth, we can form-encode the resulting parameters:
-
+For each of those `input` elements, we need to extract the `name`
+attribute and the `value` attribute, and _form-encode_ both of them.
+Form encoding is how the name/value pairs are formatted in the HTTP
+`POST` request. Basically: name, then equal sign, then value; and
+name-value pairs are separated by ampersands:
 
 ``` {.python}
-def submit_form(self, elt):
+class Tab:
+    def submit_form(self, elt):
+        # ...
+        body = ""
+        for input in inputs:
+            name = input.attributes["name"]
+            value = input.attributes.get("value", "")
+            body += "&" + name + "=" + value
+        body = body [1:]
+```
+
+Now, any time you see something like this, you've got to ask: what if
+the name or the value has an equal sign or an ampersand in it? So form
+encoding has special handling for special characters:
+
+``` {.python indent=8}
+for input in inputs:
+    # ...
+    name = percent_encode(name)
+    value = percent_encode(value)
     # ...
-    inputs = find_inputs(elt, [])
-    body = ""
-    for input in inputs:
-        name = input.attributes["name"]
-        value = input.attributes.get("value", "")
-        body += "&" + name + "=" + value.replace(" ", "%20")
-    body = body[1:]
 ```
 
-This isn't real form-encoding; this just replaces spaces by `"%20"`,
-while real form-encoding escapes characters like the equal sign, the
-ampersand, and so on. But our browser is a toy anyway, so for now
-let's just try to avoid typing equal signs, ampersands, and so on into
-forms.
+This "percent encoding" replaces all special characters with a percent
+sign followed by those characters' hex codes:
 
-Finally, we need to submit this form-encoded data in a POST request:
+``` {.python}
+def percent_encode(s):
```

Comment:
I think urlllib.quote(...) does the same thing. https://stackoverflow.com/questions/1695183/how-to-percent-encode-url-parameters-in-python

Is there extra value to doing it in the text from scratch?

Reply:
or urllib.urlencode(...)

Reply:
Well, this is a good discussion question. I generally have a preference in the book to do things from scratch. But at times I'll bend that rule, like using `makefile` for network reads to avoid looping over socket reads. Since we neither focus on percent encoding in the text nor treat it as a central concept, using urllib.quote or similar might be the right move.

Here's an alternative: what if we skip percent encoding and decoding entirely? Explain that the reader should be thinking about `&` and `=`, the answer is percent encoding, link to it, but don't implement it. It will make our web server act a little weird when used by a normal browser (spaces will become pluses?) but we can just note that in the text. The thinking here is, if we're going to do it in a library, and it's not strictly necessary, what if we don't do it?

The previous revision of this chapter had "crappy" percent encoding, and I thought that taught bad habits, but skipping it entirely seems kind of better—at least it's clear we're not even pretending and we can clearly note that this is an issue.

Reply:
I think we should keep it, since percent encoding/decoding is actually what happens.

Here are the reasons we should use urllib.quote:
* The code becomes a one-line and doesn't distract from the book contents
* It's not a good practice to implement this kind of thing yourself anyway, for security reasons (and JS has a method built-in that does it)

Reply:
Alright, I'll make a change to use a library for this.

Reply:
The new text says,

> This "percent encoding" replaces all special characters with a percent
> sign followed by those characters' hex codes; for example, a space
> becomes `%20` and a period becomes `%2e`. Python provides a
> percent-encoding function as `quote` in the `urllib` module, or you
> can write your own. (You can even skip percent encoding, but then you
> browser won't handle requests with equal signs, percent signs, or
> ampersands correctly.)